File: src/main/scala/kz/algorithms/base/Node.scala
package kz.algorithms.base

case class Node [T] (value: T, var next: Node[T] = null) {

  def addLast(value: T): Node[T] = {
    var n = this
    while (n.next!=null)
      n = n.next
    n.next = Node(value)
    this
  }

  def addFirst(value: T): Node[T] = {
    var n = Node(value)
    n.next = this
    n
  }

  override def toString = {
    this match {
      case null => ""
      case Node(v, n) => n match {
        case null => s"$v"
        case Node(v1, n1) => s"$v->$n"
      }
    }
  }

}

File: src/main/java/algorithms/Node.java
package algorithms;

public class Node {
    public Node next;
    public String value;

    public Node(String value) {
        this.value = value;
    }

    public static void main(String[] args) {
        Node root = new Node("A");
        root.next = new Node("B");
        root.next.next = new Node("C");
        root.next.next.next = new Node("D");
        root.next.next.next.next = root.next;
    }

    public static class Tree {
        Tree left;
        Tree right;
        String value;
    }

}


